src/openms/source/ANALYSIS/ID/SearchDatabase.cpp

```diff
@@ -0,0 +1,337 @@
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+// OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
+// ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// --------------------------------------------------------------------------
+// $Maintainer: Chris Bielow $
+// $Authors: Max Alcer, Heike Einsfeld $
+// --------------------------------------------------------------------------
+
+
+#include <OpenMS/ANALYSIS/ID/SearchDatabase.h>
+
+#include <OpenMS/CHEMISTRY/AAIndex.h>
+#include <OpenMS/CHEMISTRY/AASequence.h>
+#include <OpenMS/CHEMISTRY/ModificationsDB.h>
+#include <OpenMS/CHEMISTRY/ModifiedPeptideGenerator.h>
+#include <OpenMS/CHEMISTRY/ProteaseDB.h>
+#include <OpenMS/CHEMISTRY/ProteaseDigestion.h>
+#include <OpenMS/CHEMISTRY/TheoreticalSpectrumGenerator.h>
+#include <OpenMS/CONCEPT/Constants.h>
+#include <OpenMS/DATASTRUCTURES/DefaultParamHandler.h>
+#include <OpenMS/DATASTRUCTURES/Param.h>
+#include <OpenMS/FORMAT/FASTAFile.h>
+#include <OpenMS/FORMAT/MzMLFile.h>
+#include <OpenMS/KERNEL/MSExperiment.h>
+#include <OpenMS/KERNEL/Peak1D.h>
+#include <OpenMS/MATH/MISC/MathFunctions.h>
+#include <OpenMS/QC/QCBase.h>
+
+#include <omp.h>
+#include <unordered_set>
+#include <vector>
+
+using namespace std;
+
+
+namespace OpenMS
+{
+
+using namespace Constants::UserParam;
+using namespace Math;
+
+std::vector<SearchDatabase::Peptide_> SearchDatabase::generate_peptides_(const std::vector<FASTAFile::FASTAEntry>& entries) const
+{
+  vector<SearchDatabase::Peptide_> all_peptides;
+
+  ModifiedPeptideGenerator::MapToResidueType fixed_modifications = ModifiedPeptideGenerator::getModifications(modifications_fixed_);
+  ModifiedPeptideGenerator::MapToResidueType variable_modifications = ModifiedPeptideGenerator::getModifications(modifications_variable_);
+
+  #pragma omp parallel
+  { 
+    ProteaseDigestion digestor;
+    digestor.setEnzyme(digestor_enzyme_);
+    digestor.setMissedCleavages(missed_cleavages_);
+    vector<SearchDatabase::Peptide_> all_peptides_pvt;
+    #pragma omp for nowait
+    for (size_t i = 0; i < entries.size(); i++)
+    {
+      vector<AASequence> peptides;
+      digestor.digest(AASequence::fromString(entries[i].sequence), peptides, peptide_min_length_, peptide_max_length_);
+      
+      for (const auto& pep : peptides)
+      { 
+        if (pep.toString().find('X') != string::npos) continue; // filtering peptide with unknown AA, can't calculate MonoWeight
+
+        vector<AASequence> modified_peptides;
+
+        AASequence modified_pep = pep;
+
+        ModifiedPeptideGenerator::applyFixedModifications(fixed_modifications, modified_pep);
+        ModifiedPeptideGenerator::applyVariableModifications(variable_modifications, modified_pep, max_variable_mods_per_peptide_, modified_peptides);
+
+        for (const auto & mod_pep : modified_peptides)
+        {
+          double mod_seq_mz = mod_pep.getMonoWeight();
```

Comment:
maybe a bit wasteful to always sum up all masses and calculate from scratch.
but maybe not a bottleneck.

Reply:
I guess the alternative would be to adapt ModifiedPeptideGenerator to have a method that also returns the generated monomasses and adapts them incrementally.